eox_nelp/notifications/tests/test_notify_course_due_date.py

```diff
@@ -0,0 +1,138 @@
+from eox_nelp.notifications import notify_course_due_date
+from eox_nelp.notifications.models import UpcomingCourseDueDate
+from eox_nelp.notifications.notify_course_due_date import notify_upcoming_course_due_date
+from eox_nelp.tests.utils import generate_list_mock_data
+
+
+class NotifyUpcomingCourseDueDateTestCase(TestCase):
+    """Test class for function `notify_upcoming_course_due_date`."""
+
+    def setUp(self):
+        """Set common conditions for test cases."""
+        self.course_name = "course-display-name"
+        self.subsection_name = "subsection-display-name"
+        self.recipients_list = ["vader@star.com", "palpatine@star.com"]
+        self.subsection_due_date = timezone.now() + timedelta(days=4)
+        self.enrollments_data = [
+            {"user": {"email": self.recipients_list[0]}},
+            {"user": {"email": self.recipients_list[1]}},
+        ]
+        self.subsection_xblock_data = [
+            {"display_name": self.subsection_name, "due": self.subsection_due_date},
+        ]
+        self.course_xblock_data = [
+            {
+                "display_name": self.course_name,
+            },
+        ]
+
+    @patch("eox_nelp.notifications.notify_course_due_date.send_email_multialternative")
+    @patch("eox_nelp.notifications.notify_course_due_date.CourseEmailTemplate")
+    @patch("eox_nelp.notifications.notify_course_due_date.get_xblock_course_from_usage_key_string")
+    @patch("eox_nelp.notifications.notify_course_due_date.get_xblock_from_usage_key_string")
+    @patch("eox_nelp.notifications.notify_course_due_date.CourseEnrollment")
+    def test_notify_upcoming_course_due_date(  # pylint: disable=too-many-arguments
+        self,
+        course_enrollment_mock,
+        get_xblock_from_usage_key_string_mock,
+        get_xblock_course_from_usage_key_string_mock,
+        course_email_template_mock,
+        send_email_multialternative_mock,
+    ):
+        """Test correct behaviour of sending email with
+        method `notify_upcoming_course_due_date`.
+        Expected behavior:
+            - email_key_words are presented in the `text_message` arg (0) of 2nd call of course_email_template.
+            - email_key_words are presented in the `html_message` arg (0) of 3rd call of course_email_template.
+            - course_name is presented in `subject` arg(0) of the call of send_email_multialternative.
+            - subsection_name is presented in `subject` arg(0) of the call of send_email_multialternative.
+            - recipient list is presented in `recipient_emails` arg(3) of the call of send_email_multialternative.
+            - The email was sent from the correct email address (if specified)
+            - sent attribute of instance of `UpcomingCourseDueDate` after the process is True.
+            - Log successful of sent message.
+        """
+        UpcomingCourseDueDate.objects.all().delete()  # pylint: disable=no-member
```

Comment:
why do you need this?

Reply:
Because I am creating the same instance of UpcomingCourseDueDate, with the same data. In the first one I pass the sent to true, in the second verify the False. So it could change depending the order or running functions. To avoid that, I ensure remove all. Othe option is to create and instance with different data.

Reply:
`get_or_create` or `update_or_create ` because that makes me think that your test only works with an empty database

Reply:
[cdd2136](https://github.com/eduNEXT/eox-nelp/pull/44/commits/cdd2136830dd5425dd813288ac834c2add3190c0)